compiler/rustc_resolve/src/errors.rs

```diff
@@ -0,0 +1,279 @@
+    pub(crate) sub_reachable: Option<LabelWithSimilarNameReachable>,
+
+    #[subdiagnostic]
+    pub(crate) sub_reachable_suggestion: Option<TryUsingSimilarlyNamedLabel>,
+
+    #[subdiagnostic]
+    pub(crate) sub_unreachable: Option<UnreachableLabelWithSimilarNameExists>,
+}
+
+#[derive(SessionSubdiagnostic)]
+#[label(resolve::label_with_similar_name_reachable)]
+pub(crate) struct LabelWithSimilarNameReachable(#[primary_span] pub(crate) Span);
+
+#[derive(SessionSubdiagnostic)]
+#[suggestion(
+    resolve::try_using_similarly_named_label,
+    code = "{ident_name}",
+    applicability = "maybe-incorrect"
+)]
+pub(crate) struct TryUsingSimilarlyNamedLabel {
+    #[primary_span]
+    pub(crate) span: Span,
+
+    pub(crate) ident_name: Symbol,
+}
+
+#[derive(SessionSubdiagnostic)]
+#[label(resolve::unreachable_label_with_similar_name_exists)]
+pub(crate) struct UnreachableLabelWithSimilarNameExists {
+    #[primary_span]
+    #[label]
+    pub(crate) ident_span: Span,
+}
+
+#[derive(SessionDiagnostic)]
+#[diag(resolve::self_import_can_only_appear_once_in_the_list, code = "E0430")]
+pub(crate) struct SelfImportCanOnlyAppearOnceInTheList {
+    #[primary_span]
+    #[label]
+    pub(crate) span: Span,
+}
+
+#[derive(SessionDiagnostic)]
+#[diag(resolve::self_import_only_in_import_list_with_non_empty_prefix, code = "E0431")]
+pub(crate) struct SelfImportOnlyInImportListWithNonEmptyPrefix {
+    #[primary_span]
+    #[label]
+    pub(crate) span: Span,
+}
+
+#[derive(SessionDiagnostic)]
+#[diag(resolve::cannot_capture_dynamic_environment_in_fn_item, code = "E0434")]
+pub(crate) struct CannotCaptureDynamicEnvironmentInFnItem<'a> {
+    #[primary_span]
+    #[help]
+    pub(crate) span: Span,
+
+    pub(crate) closure_hint: &'a str,
+}
+
+#[derive(SessionDiagnostic)]
+#[diag(resolve::attempt_to_use_non_constant_value_in_constant, code = "E0435")]
+pub(crate) struct AttemptToUseNonConstantValueInConstant<'a> {
+    #[primary_span]
+    #[help]
+    pub(crate) span: Span,
+
+    #[subdiagnostic]
+    pub(crate) with: Option<AttemptToUseNonConstantValueInConstantWithSuggestion<'a>>,
+
+    #[subdiagnostic]
+    pub(crate) with_label: Option<AttemptToUseNonConstantValueInConstantWithSuggestionLabel>,
+
+    #[subdiagnostic]
+    pub(crate) without: Option<AttemptToUseNonConstantValueInConstantWithoutSuggestion<'a>>,
+}
+
+#[derive(SessionSubdiagnostic)]
+#[suggestion(
+    resolve::attempt_to_use_non_constant_value_in_constant_with_suggestion,
+    code = "{suggestion} {ident}",
+    applicability = "maybe-incorrect"
+)]
+pub(crate) struct AttemptToUseNonConstantValueInConstantWithSuggestion<'a> {
+    #[primary_span]
+    pub(crate) span: Span,
+
+    pub(crate) ident: Ident,
+    pub(crate) suggestion: &'a str,
+    pub(crate) current: &'a str,
+}
+
+#[derive(SessionSubdiagnostic)]
+#[label(resolve::attempt_to_use_non_constant_value_in_constant_with_suggestion_label)]
+pub(crate) struct AttemptToUseNonConstantValueInConstantWithSuggestionLabel {
```

Comment:
I think you've got some of these types mixed up with the Fluent messages, it looks like this message expects a `suggestion` field that it doesn't have but that `AttemptToUseNonConstantValueInConstantWithoutSuggestion` does (despite its name).

Reply:
tried to rename this type to more meaningful and understandable as per context.